### ISSUE TITLE
Long headers in Excel file crash the Excel UI

### ISSUE THREAD

**Issue by castillohair** [taborlab/FlowCal#276]
As title says. In my case, a header such as "520nm Light Intensity (µmol/(m^2*s))" seems to crash the Excel UI. The problem seems to be related to the regex that we use to extract channel names from headers, as they seem to be incurring in [catastrophic backtracking](http://www.regular-expressions.info/catastrophic.html).

Because of my current time availability, I will not be able to provide more details until tomorrow.

**Comment by castillohair** [comment]
(Let's ignore the fact that I said "tomorrow" and it's been almost two months)

The problem is definitely, as I said, a case of [catastrophic backtracking](https://www.regular-expressions.info/catastrophic.html). I'll offer a short explanation below, which assumes some familiarity with how the regex engine attempts to match strings. For more information on this, refer to this [regex tutorial](https://www.regular-expressions.info/tutorialcnt.html) or any other regex reference.

Our regex for the "Units" column is currently as follows:
```
^\s*(\S(.*\S)*)\s*Units\s*$
```

With a string such as `520nm Light Intensity (umol/(m^2*s))`, the regex engine has several ways in which it can attempt to match the first portion of the regex pattern (before `\s*Units\s*`). A few expansions of the quantifiers, in the order that the regex engine tries them (as far as I understand) are as follows:

```
"520nm Light Intensity (umol/(m^2*s))"
 S...................................
 S..................................S

 S.................................S.
 S.................................SS

 S................................S..
 S................................S.S
 S................................SS.
 S................................SSS

 S...............................S...
 S...............................S..S
 S...............................S.S.
 S...............................S.SS
 S...............................SS..
 S...............................SS.S
 S...............................SSS.
 S...............................SSSS
```

I will be referring to `"520nm Light Intensity (umol/(m^2*s))"` as the "string", and to each one of my expanded regex patterns as "pattern", starting from the one immediately below the string. The first `S` on each row corresponds to the first `\S` inside the first group (set of parenthesis), before the opening of the second group. Then each sequence of dots corresponds to the `.*` inside the second group, and a `S` at the end corresponds to the `\S` at the end of `(.*\S)`.

A regex engine tries to match as many dots as possible, and when it fails, it will *backtrack* and match the previous number of characters except for the last one. This is what it does in the first pattern in my example. At the end of the first block (second pattern), the engine has successfully matched one repetition of `(.*\S)`, but the whole regex has not matched, so the engine backtracks again.

In the next block (patterns 3 and 4), the first `(.*\S)` matches all characters until one before the end of the string. But because there is a `*` at the end of the second group, the remaining character of the string is matched by a second repetition of `(.*\S)`. In pattern four, this second repetition has matched successfully, but the whole regex pattern has not.

In the next block (patterns 5-8), the first `(.*\S)` matches all characters until two before the end of the string. The engine then attempts to match these two characters to one (patterns 5 and 6) or two (patterns 7 and 8) repetitions of group 2.

In the final block that I illustrated, the regex engine matches the first `(.*\S)` to all characters until three before the end of the string, and the remaining characters are matched by either one (patterns 9 and 10), two (patterns 11-14) or three (patterns 15 and 16) additional repetitions of group 2 `(.*\S)`.

Following this reasoning, the number of possibilities that the regex must attempt grows **exponentially** with the length of the string. This is referred to as *catastrophic backtracking*, and it happens because most characters in the string are matched both by `\S` and `.`, and because we have nested groups with quantifiers. These two features give rise to exponential time growth.

To summarize, catastrophic backtracking has the following features:
- Result of poorly designed regex pattern, in which regex tokens that match similar sets of characters are combined with quantifiers and nested groups.
- Run time grows exponentially with the length of the string.
- It may not be completely obvious from my example, but catastrophic backtracking does **not** manifest itself when the string matches. It is only when the string doesn't match that a ridiculously large number of possibilities must be attempted for the engine to "convince itself" that a match is not possible.

**Comment by castillohair** [comment]
For further evidence that we are dealing with catastrophic backtracking, I made the following script:

```python
import re
import time

# Same as currently in excel_ui
r = re.compile(r'^\s*(\S(.*\S)*)\s*Units\s*$')

# String that triggers the problem for me
s = '520nm Light Intensity (umol/(m^2*s))'

for i in range(10, len(s)):
    # Measure elapsed time
    ts = time.time()
    m = r.match(s[:i])
    dt = time.time() - ts
    # Print results
    print("For string \"{}\" (length {}), time was {:.5f} s.".format(s[:i], i, dt))
    # The following should never work
    if m is not None:
        print("Matched!")
```

Note that the regex is the same as we currently have in `develop` in the excel UI. The output of the program is as follows:

```
For string "520nm Ligh" (length 10), time was 0.00000 s.
For string "520nm Light" (length 11), time was 0.00000 s.
For string "520nm Light " (length 12), time was 0.00000 s.
For string "520nm Light I" (length 13), time was 0.00000 s.
For string "520nm Light In" (length 14), time was 0.00000 s.
For string "520nm Light Int" (length 15), time was 0.00000 s.
For string "520nm Light Inte" (length 16), time was 0.00000 s.
For string "520nm Light Inten" (length 17), time was 0.01500 s.
For string "520nm Light Intens" (length 18), time was 0.00000 s.
For string "520nm Light Intensi" (length 19), time was 0.01600 s.
For string "520nm Light Intensit" (length 20), time was 0.03100 s.
For string "520nm Light Intensity" (length 21), time was 0.04700 s.
For string "520nm Light Intensity " (length 22), time was 0.06300 s.
For string "520nm Light Intensity (" (length 23), time was 0.10900 s.
For string "520nm Light Intensity (u" (length 24), time was 0.20300 s.
For string "520nm Light Intensity (um" (length 25), time was 0.42300 s.
For string "520nm Light Intensity (umo" (length 26), time was 0.84200 s.
For string "520nm Light Intensity (umol" (length 27), time was 1.71400 s.
For string "520nm Light Intensity (umol/" (length 28), time was 3.30100 s.
For string "520nm Light Intensity (umol/(" (length 29), time was 6.67700 s.
For string "520nm Light Intensity (umol/(m" (length 30), time was 13.38800 s.
For string "520nm Light Intensity (umol/(m^" (length 31), time was 26.74700 s.
For string "520nm Light Intensity (umol/(m^2" (length 32), time was 53.58100 s.
For string "520nm Light Intensity (umol/(m^2*" (length 33), time was 107.38100 s.
For string "520nm Light Intensity (umol/(m^2*s" (length 34), time was 220.25300 s.
For string "520nm Light Intensity (umol/(m^2*s)" (length 35), time was 436.52400 s.
```

We can clearly see the exponential time growth.

To show that this is not a problem exclusive to my string, I replaced it with the following text: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.`. The following is the resulting output:

```
For string "Lorem ipsu" (length 10), time was 0.00000 s.
For string "Lorem ipsum" (length 11), time was 0.00000 s.
For string "Lorem ipsum " (length 12), time was 0.00000 s.
For string "Lorem ipsum d" (length 13), time was 0.00000 s.
For string "Lorem ipsum do" (length 14), time was 0.00000 s.
For string "Lorem ipsum dol" (length 15), time was 0.00000 s.
For string "Lorem ipsum dolo" (length 16), time was 0.00000 s.
For string "Lorem ipsum dolor" (length 17), time was 0.00000 s.
For string "Lorem ipsum dolor " (length 18), time was 0.00000 s.
For string "Lorem ipsum dolor s" (length 19), time was 0.01600 s.
For string "Lorem ipsum dolor si" (length 20), time was 0.01500 s.
For string "Lorem ipsum dolor sit" (length 21), time was 0.01600 s.
For string "Lorem ipsum dolor sit " (length 22), time was 0.03100 s.
For string "Lorem ipsum dolor sit a" (length 23), time was 0.04700 s.
For string "Lorem ipsum dolor sit am" (length 24), time was 0.09800 s.
For string "Lorem ipsum dolor sit ame" (length 25), time was 0.21800 s.
For string "Lorem ipsum dolor sit amet" (length 26), time was 0.45700 s.
For string "Lorem ipsum dolor sit amet," (length 27), time was 0.87700 s.
For string "Lorem ipsum dolor sit amet, " (length 28), time was 1.08000 s.
For string "Lorem ipsum dolor sit amet, c" (length 29), time was 1.97700 s.
For string "Lorem ipsum dolor sit amet, co" (length 30), time was 3.73300 s.
For string "Lorem ipsum dolor sit amet, con" (length 31), time was 7.16200 s.
For string "Lorem ipsum dolor sit amet, cons" (length 32), time was 14.05400 s.
For string "Lorem ipsum dolor sit amet, conse" (length 33), time was 27.34700 s.
For string "Lorem ipsum dolor sit amet, consec" (length 34), time was 56.00000 s.
For string "Lorem ipsum dolor sit amet, consect" (length 35), time was 111.93300 s.
For string "Lorem ipsum dolor sit amet, consecte" (length 36), time was 219.59700 s.
For string "Lorem ipsum dolor sit amet, consectet" (length 37), time was 412.87600 s.
For string "Lorem ipsum dolor sit amet, consectetu" (length 38), time was 805.42500 s.
For string "Lorem ipsum dolor sit amet, consectetur" (length 39), time was 1609.98900 s.
Traceback (most recent call last):
  File ".\regex_test_2.py", line 13, in <module>
    m = r.match(s[:i])
KeyboardInterrupt
```

(I stopped the program after almost an hour).

Finally, to show that this is not an issue when the string matches, I modified the script to always add `" Units"` to the string:

```python
import re
import time

# Same as currently in excel_ui
r = re.compile(r'^\s*(\S(.*\S)*)\s*Units\s*$')

# String that triggers the problem for me
s = '520nm Light Intensity (umol/(m^2*s))'

for i in range(10, len(s)):
    si = s[:i] + " Units"
    # Measure elapsed time
    ts = time.time()
    m = r.match(si)
    dt = time.time() - ts
    # Print results
    print("For string \"{}\" (length {}), time was {:.5f} s.".format(si, i, dt))
    # The following should always work
    if m is not None:
        print("Matched!")
```

The program run almost immediately, and the output was as follows:

```
For string "520nm Ligh Units" (length 10), time was 0.00000 s.
Matched!
For string "520nm Light Units" (length 11), time was 0.00000 s.
Matched!
For string "520nm Light  Units" (length 12), time was 0.00000 s.
Matched!
For string "520nm Light I Units" (length 13), time was 0.00000 s.
Matched!
For string "520nm Light In Units" (length 14), time was 0.00000 s.
Matched!
For string "520nm Light Int Units" (length 15), time was 0.00000 s.
Matched!
For string "520nm Light Inte Units" (length 16), time was 0.00000 s.
Matched!
For string "520nm Light Inten Units" (length 17), time was 0.00000 s.
Matched!
For string "520nm Light Intens Units" (length 18), time was 0.00000 s.
Matched!
For string "520nm Light Intensi Units" (length 19), time was 0.00000 s.
Matched!
For string "520nm Light Intensit Units" (length 20), time was 0.00000 s.
Matched!
For string "520nm Light Intensity Units" (length 21), time was 0.00000 s.
Matched!
For string "520nm Light Intensity  Units" (length 22), time was 0.00000 s.
Matched!
For string "520nm Light Intensity ( Units" (length 23), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (u Units" (length 24), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (um Units" (length 25), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umo Units" (length 26), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol Units" (length 27), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/ Units" (length 28), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/( Units" (length 29), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m Units" (length 30), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^ Units" (length 31), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2 Units" (length 32), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2* Units" (length 33), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2*s Units" (length 34), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2*s) Units" (length 35), time was 0.00000 s.
Matched!
```

**Comment by castillohair** [comment]
The obvious solution is to remove the second group, which results in the following regex:

```
^\s*(\S.*\S)\s*Units\s*$
```

However, as indicated in #251, this misses single-character channel names.

The reason why we need the second group with a `*` quantifier is because we want this to be optional (zero repetitions), and not necessarily because we need lots of repetitions of this group. The variable length of a channel name is already taken care of by the `.*` portion. Furthermore, unlimited repetition of the second group is what causes the catastrophic backtracking. Therefore, the following is a feasible solution:

```
^\s*(\S(.*\S)?)\s*Units\s*$
```

Where a `?` quantifier replaces the `*` after the second group. This makes this group optional, as in it can match once or not match.

I would also propose to make the second group a [non-capturing group](https://www.regular-expressions.info/brackets.html). The only group that we need exposed is the outer first group. Finally, I would argue that there should be at least one space between the channel name and the word "Units". This would make strings like "FL1Units" not match, which I think is ok.

With these modifications, the following regex results:

```
^\s*(\S(?:.*\S)?)\s+Units\s*$
```

If we use this regex in out previous script:

```python
import re
import time

# New proposed regex
r = re.compile(r'^\s*(\S(?:.*\S)?)\s+Units\s*$')

# String that triggers the problem for me
s = '520nm Light Intensity (umol/(m^2*s))'

for i in range(10, len(s)):
    # Measure elapsed time
    ts = time.time()
    m = r.match(s[:i])
    dt = time.time() - ts
    # Print results
    print("For string \"{}\" (length {}), time was {:.5f} s.".format(s[:i], i, dt))
    # The following should never work
    if m is not None:
        print("Matched!")
```

This script runs almost instantly, with the following output:

```
For string "520nm Ligh" (length 10), time was 0.00000 s.
For string "520nm Light" (length 11), time was 0.00000 s.
For string "520nm Light " (length 12), time was 0.00000 s.
For string "520nm Light I" (length 13), time was 0.00000 s.
For string "520nm Light In" (length 14), time was 0.00000 s.
For string "520nm Light Int" (length 15), time was 0.00000 s.
For string "520nm Light Inte" (length 16), time was 0.00000 s.
For string "520nm Light Inten" (length 17), time was 0.00000 s.
For string "520nm Light Intens" (length 18), time was 0.00000 s.
For string "520nm Light Intensi" (length 19), time was 0.00000 s.
For string "520nm Light Intensit" (length 20), time was 0.00000 s.
For string "520nm Light Intensity" (length 21), time was 0.00000 s.
For string "520nm Light Intensity " (length 22), time was 0.00000 s.
For string "520nm Light Intensity (" (length 23), time was 0.00000 s.
For string "520nm Light Intensity (u" (length 24), time was 0.00000 s.
For string "520nm Light Intensity (um" (length 25), time was 0.00000 s.
For string "520nm Light Intensity (umo" (length 26), time was 0.00000 s.
For string "520nm Light Intensity (umol" (length 27), time was 0.00000 s.
For string "520nm Light Intensity (umol/" (length 28), time was 0.00000 s.
For string "520nm Light Intensity (umol/(" (length 29), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m" (length 30), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m^" (length 31), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m^2" (length 32), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m^2*" (length 33), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m^2*s" (length 34), time was 0.00000 s.
For string "520nm Light Intensity (umol/(m^2*s)" (length 35), time was 0.00000 s.
```

The same results with the string "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.". Finally, with the following script:

```python
import re
import time

# New proposed regex
r = re.compile(r'^\s*(\S(?:.*\S)?)\s+Units\s*$')

# String that triggers the problem for me
s = '520nm Light Intensity (umol/(m^2*s))'

for i in range(10, len(s)):
    si = s[:i] + " Units"
    # Measure elapsed time
    ts = time.time()
    m = r.match(si)
    dt = time.time() - ts
    # Print results
    print("For string \"{}\" (length {}), time was {:.5f} s.".format(si, i, dt))
    # The following should always work
    if m is not None:
        print("Matched!")
```

The following results almost instantly:

```
For string "520nm Ligh Units" (length 10), time was 0.00000 s.
Matched!
For string "520nm Light Units" (length 11), time was 0.00000 s.
Matched!
For string "520nm Light  Units" (length 12), time was 0.00000 s.
Matched!
For string "520nm Light I Units" (length 13), time was 0.00000 s.
Matched!
For string "520nm Light In Units" (length 14), time was 0.00000 s.
Matched!
For string "520nm Light Int Units" (length 15), time was 0.00000 s.
Matched!
For string "520nm Light Inte Units" (length 16), time was 0.00000 s.
Matched!
For string "520nm Light Inten Units" (length 17), time was 0.00000 s.
Matched!
For string "520nm Light Intens Units" (length 18), time was 0.00000 s.
Matched!
For string "520nm Light Intensi Units" (length 19), time was 0.00000 s.
Matched!
For string "520nm Light Intensit Units" (length 20), time was 0.00000 s.
Matched!
For string "520nm Light Intensity Units" (length 21), time was 0.00000 s.
Matched!
For string "520nm Light Intensity  Units" (length 22), time was 0.00000 s.
Matched!
For string "520nm Light Intensity ( Units" (length 23), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (u Units" (length 24), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (um Units" (length 25), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umo Units" (length 26), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol Units" (length 27), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/ Units" (length 28), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/( Units" (length 29), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m Units" (length 30), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^ Units" (length 31), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2 Units" (length 32), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2* Units" (length 33), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2*s Units" (length 34), time was 0.00000 s.
Matched!
For string "520nm Light Intensity (umol/(m^2*s) Units" (length 35), time was 0.00000 s.
Matched!
```

Therefore, I think the modified regular expression solves this issue. I will now submit a pull request with more formal unit testing.

**Comment by JS3xton** [comment]
This is fantastic. Excellent explanation. Good, precise solution.

> (Let's ignore the fact that I said "tomorrow" and it's been almost two months)

😂

> Result of poorly designed regex pattern

😬


`*` -> `?` for inner group = 👍
non-capturing inner group = 👍
space before "Units" I feel less strongly about, but am fine with what you have here

**Comment by castillohair** [comment]
Forgot to document how the new tests perform with the old regex.

6f4993caa390c35486df9d044347f2771dddea7e contains unit tests but not the changes discussed above. The results of the unit test for the Units regex are as follows:

```
$ python -m unittest test.test_excel_ui.TestRegexUnits
F
======================================================================
FAIL: test_match (test.test_excel_ui.TestRegexUnits)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_excel_ui.py", line 323, in test_match
    self.assertIsNone(m)
AssertionError: <_sre.SRE_Match object at 0x7fc06e38ce88> is not None

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
```

This is due to test string `Parameter 14Units`, which should not match under the new regex due to the lack of space before `Units`, but actually matches using the old regex and therefore the test fails. If I remove this test, the string `'520nm Light Intensity (umol/(m^2*s))'` is attempted to be matched and FlowCal enters into catastrophic backtracking (program stays in a long loop and doesn't respond).

**Comment by castillohair** [comment]
Pull request merged. Feature should be in the next FlowCal release. Will close issue.
